sky/engine/core/script/dart_controller.cc

```diff
@@ -68,6 +68,16 @@ DartController::~DartController() {
   if (dom_dart_state_) {
     // Don't use a DartIsolateScope here since we never exit the isolate.
     Dart_EnterIsolate(dom_dart_state_->isolate());
+
+    // Run the framework's shutdown handler.
+    {
+      DartApiScope dart_api_scope;
+      Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
+      if (!LogIfError(ui_library)) {
+        DartInvokeField(ui_library, "_onShutdown", {});
+      }
+    }
```

Comment:
I'm a little curious what we intend to use this callback for.  In the
general case we might not be able to promise that callbacks run before
shutdown occurs (say if the OS kills our process).

Is the Dart isolate holding onto any heavyweight state that needs to be
freed sooner than it might otherwise?

I am commenting because this change might be indicating a demand for some
more formal lifecycle management perhaps.

On Fri, Nov 13, 2015, 6:07 PM Adam Barth <notifications@github.com> wrote:

> In sky/engine/core/script/dart_controller.cc
> <https://github.com/flutter/engine/pull/2051#discussion_r44852995>:
>
> > @@ -68,6 +68,16 @@ DartController::~DartController() {
> >    if (dom_dart_state_) {
> >      // Don't use a DartIsolateScope here since we never exit the isolate.
> >      Dart_EnterIsolate(dom_dart_state_->isolate());
> > +
> > +    // Run the framework's shutdown handler.
> > +    {
> > +      DartApiScope dart_api_scope;
> > +      Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
> > +      if (!LogIfError(ui_library)) {
> > +        DartInvokeField(ui_library, "_onShutdown", {});
> > +      }
> > +    }
>
> It's scary to run this from inside the destructor. Can we make shutdown
> more explicit?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flutter/engine/pull/2051/files#r44852995>.
>